# Prediction a individual game result in the NBA

## Business Understanding
 Pequena intro sobre nba? e apostas?
per email do prof -> Motivation (what is the problem you are trying to solve? If not obvious, please also
 add why this problem is important) 
 
### Data Mining Goals and Success Criteria
With this project we aim to be able to predict the winner of a given match based on previous games.


## Data Understanding

For this project the data sets used were obtaing from the Kaggle data base ( https://www.kaggle.com/pablote/nba-enhanced-stats )
It contains 17 files described in table 1. 

| Name | Size | File | Description |
|---|---|---|---|
| 2012-18_officialBoxScore | 25,5 MB | '.csv ' | |
| 2012-18_playerBoxScore | 35,2 MB | '.csv ' | |
| 2012-18_standings | 5,34 MB | '.csv ' | |
| 2012-18_teamBoxScore | 8,88 MB | '.csv ' | |
| 2016-17_officialBoxScore | 4,26 MB | '.csv ' | |
| 2016-17_playerBoxScore | 5,93 MB | '.csv ' | |
| 2016-17_standings | 892 KB | '.csv ' | |
| 2016-17_teamBoxScore | 1,48 MB | '.csv ' | |
| 2017-18_officialBoxScore | 4,26 MB | '.csv ' | |
| 2017-18_playerBoxScore | 5,92 MB | '.csv ' | |
| 2017-18_standings | 936 KB | '.csv ' | |
| 2017-18_teamBoxScore | 1,48 MB | '.csv ' | |
| metadata_officialBoxScore | 42 KB | '.pdf ' | |
| metadata_playerBoxScore | 35 KB | '.pdf ' | |
| metadata_standing | 34 KB | '.pdf ' | |
| metadata_teamBoxScore | 42 KB | '.pdf ' | |
| teamBoxScore | 714 KB | '.csv ' | |
<caption> **Table 1 - Data sets used** </caption> 

From these, we selected only the ones with data from 2012 to 2018 since they gather more information.
Next we started describing and exploring the four selected dataframes.


In [14]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from datetime import datetime

In [15]:
pd.set_option('display.max_columns', None)
#pd.set_option('display.max_rows', 50)

In [16]:
#Import data
standings = pd.read_csv('2012-18_standings.csv')
boxscore = pd.read_csv('2012-18_officialBoxScore.csv')
player = pd.read_csv('2012-18_playerBoxScore.csv')
team = pd.read_csv('2012-18_teamBoxScore.csv')

### Data description, exploration and quality verification

#### Standings

In [17]:
#exploring Standings
standings.index
standings.columns

featstandings = list(standings) 
standings.isnull().any(axis = 1).sum()
standings.dtypes


stDate             object
teamAbbr           object
rank                int64
rankOrd            object
gameWon             int64
gameLost            int64
stk                object
stkType            object
stkTot              int64
gameBack          float64
ptsFor              int64
ptsAgnst            int64
homeWin             int64
homeLoss            int64
awayWin             int64
awayLoss            int64
confWin             int64
confLoss            int64
lastFive            int64
lastTen             int64
gamePlay            int64
ptsScore          float64
ptsAllow          float64
ptsDiff           float64
opptGmPlay          int64
opptGmWon           int64
opptOpptGmPlay      int64
opptOpptGmWon       int64
sos               float64
rel%Indx          float64
mov               float64
srs               float64
pw%               float64
pyth%13.91        float64
wpyth13.91        float64
lpyth13.91        float64
pyth%16.5         float64
wpyth16.5         float64
lpyth16.5   

Standings is a dataframe with 34 variables and 29520 events, and zero missing values.
Variable are: 
- stDate: standing date
- teamAbbr: team name abbreviation
- rank: Ranking
- rankOrd: Ranking Position
- gameWon: if the game was won
- gameLost: if the game was lost
- stk: streak
- stkType: Wether the streak is of losses or wins
- stkTot:
- gameBack: games back from first place
- ptsFor: Points Scores
- ptsAgnst: Points allowed
- homeWin: Home wins
- hoeLoss: Home losses
- awayWin: Away wins
- awayLoss: Aways losses
- confWin: Wins against same conference
- confLoss: Losses against same conference
- lastFive: Results in last 5
- lastTen: Resultas in last 10
- gamePlay: Games played
- ptsScore: Points scores
- ptsAllow: Points allowed
- ptsDiff: Point differential
- opptGmPlay: Opponents Games played
- opptGmWon: Opponents Game won
- opptOpptGmPlay:
- opptOpptGmWon:
- sos: strenght of schedule
- rel%Indx:
- mov:
- srs:
- pw%:
- pyth%13.91: pythagorean win percentage
- wpyth13.91: pythagorean wins
- lpyth13.91: pythagorens losses
- pyth%16.5: pythagorean win percentage
- wpyth16.5: pythagorean wins
- lpyth16.5: pythagorean losses

#### Boxscore

In [18]:
#exploring boxscore
boxscore.index
boxscore.describe()
boxscore.columns
columns = list(boxscore) 
featboxscore = list(boxscore)
boxscore.isnull().any(axis = 1).sum()
boxscore.dtypes

gmDate         object
gmTime         object
seasTyp        object
offLNm         object
offFNm         object
               ...   
opptAR        float64
opptAST/TO    float64
opptSTL/TO    float64
poss          float64
pace          float64
Length: 119, dtype: object

BoxScore is a dataframe with 119 variables and 44284 events, and zero missing values. Variable are: 
- gmDate date of game
- gmTime: starting time of game
- seasTyp: playoffs or regular season
- offLNm: the 3 officials last name
- offFNm: the 3 officials first name
- teamAbbr: team name
- teamConf: team conference
- teamDiv: team division
- teamLoc: team location
- teamRslt: team result
- teamMin: teams minutes
- teamDayOff: teams days of rest
- teamPTS: team points
- teamAST: team assists
- teamTO: team turnovers
- teamSTL: teams steals
- teamBLK: team blocks
- teamPF: team personal fouls
- teamFGA: team field goal attempter
- teamFGM: team field goal made
- teamFG%: team field goal percentage
- team2PA: team 2 points attempted
- team2PM: team 2 points made
- team2P%: team 2 points percentage
- team3PA: team three points attempted
- team3PM: team three points made
- team3P%: team three point percentage
- teamFTA: team free throws attempted
- teamFTM: team free throws made
- teamFT%: team free throw percentage
- teamORB: team offensive rebounds
- teamDRB: team defensive rebounds
- teamTRB: team total rebounds
- teamPTS1: team points in the first quarter
- teamPTS2: team points in the second quarter
- teamPTS3: team points in the third quarter
- teamPTS4: team points in the fourth quarter
- teamPTS5: team points in the first overtime
- teamPTS6: team points in the second overtime
- teamPTS7: team points in third overtime
- teamPTS8: team points in fourth overtime
- teamTREB%: team total rebound percentage
- teamASST%: team assist percentage
- teamTS%: team true shooting percentage
- teamEFG%: team effective field goal percentage
- teamOREB%: team  offensive rebound percentage
- teamDREB%: team defensive rebound percentage
- teamTO%: team turnover percentage
- teamSTL%: team steal percentage
- teamBLK%: team block percentage
- teamBLKR: team block rate
- teamPPS: team point per shot
- teamFIC: team floor impact counter
- teamFIC40: 
- teamOrtg: team offensive rating
- teamDrtg: team defensive rating
- teamEDiff:
- teamPlay%:
- teamAR:
- teamAST/TO: team assists per turnover ratio
- teamSTL/TO: team steal per turnover ratio
- opptAbbr: opponent name
- opptConf: opponent conference
- opptDiv: opponent division
- opptLoc: wether the opponent is home or away
- opptRslt: opponent result
- opptMin: opponents minutes
- opptDayOff: opponents day off before game
- opptPTS: opponents points
- opptAST: opponents assists
- opptTO: opponents turnovers
- opptSTL: opponents steals
- opptBLK: opponents blocks
- opptPF: opponents personal fouls
- opptFGA: opponents field goals attempter
- opptFGM: opponents field goals made
- opptFG%: opponents field goal percentage
- oppt2PA: opponents 2 pointers attempter
- oppt2PM: opponents 2 pointers made
- oppt2P%: opponents 2 point percentage
- oppt3PA: opponents 3 pointers attempted
- oppt3PM: opponents 3 pointers made
- oppt3P%: opponents 3 point percentage
- opptFTA: opponents free throw attempted
- opptFTM: opponents free throw made
- opptFT%: opponents free throw percentage
- opptORB: opponents ofensive rebounds
- opptDRB: opopnents defensive rebounds
- opptTRB: opponents total rebounds
- opptPTS1: opponents points in the first quarter
- opptPTS2: opponents points in the second quarter
- opptPTS3: opponents points in the third quarter
- opptPTS4: opponents points in the fourth quarter
- opptPTS5: oppoentns points in first overtime
- opptPTS6: opponents points in the second overtime
- opptPTS7: opponents points in the third overtime
- opptPTS8: opponents points in the fourth overtime
- opptTREB%: opponent total rebound percentage
- opptASST%: opponent assist percentage
- opptTS%: opponent true shooting percentage
- opptEFG%: opponent effective field goal percentage
- opptOREB%: opponent offensive rebound percentage
- opptDREB%: opponent defensive rebound percentage
- opptTO%: opponent turnover percentage
- opptSTL%: opponent steal percentage
- opptBLK%: opponent block percentage
- opptBLKR: opponent block rate
- opptPPS: opponent point per shot
- opptFIC: opponent floor impact counter
- opptFIC40:
- opptOrtg- opponent offensive rating
- opptDrtg- opponent defensive rating
- opptEDiff:
- opptPlay%:
- opptAR:
- opptAST/TO: opponent assist per turnover ratio
- opptSTL/TO: opponent steal per turnover ratio
- poss: possesions per game
- pace: pace of the game


#### Player

In [19]:
#exploring player
player.index 
player.describe() 
player.columns 
columns = list(player) 
columns
len(columns)
featplayer = list(player)
player.isnull().any(axis = 1).sum()
player = player.dropna(axis = 0, how='any')
player = player.reset_index(drop=True)


Player is a dataframe with 51 variables and 155713 events, and 41 rows with missing values. Variable are: 
- gmDate: Game date
- gmTime: starting time of the game
- seasTyp: Playoffs or regular season
- playLNm: players last name
- playFNm: players first name
- teamAbbr: Team
- teamConf: Team conference
- teamDiv: Team division
- teamLoc: Team location
- teamRslt: Team result
- teamDayOff: Teams day off rest
- offLNm1: First offical last name
- offFNm1: First official first name
- offLNm2: Second official last name
- offFNm2: Second official first name
- offLNm3: Third official last name
- offFNm3: Third official firt name
- playDispNm: Player shirt name
- playStat: Starter or off the bench player
- playMin: minutes played
- playPos: Player Position
- playHeight: Player Height
- playWeight: Player Weight
- playBDate: Player birthdate
- playPTS: Player Points
- playAST: Player Assists
- playTO: Player turnovers
- playSTL: Player steals
- playBLK: Player Blocks
- playPF: Player Personal Fouls
- playFGA: Player Field Goals Attempted
- playFGM: Player Field Goals Made
- playFG%: Player Field Goal Percentage
- play2PA: Player 2 points attempted
- play2PM: Player 2 points made
- play2P%: PLayer 2 points percentage
- play3PA: Player 3 points attempted
- play3PM: Player 3 points made
- play3P%: Player 3 points percentage
- playFTA: Player Free Throws attempted
- playFTM: Player Free Throws made
- playFT%: PLayer Free Throws percentage
- playORB: Player Offensive Rebounds
- playDRB: Player Defensive Rebounds
- playTRB: Player Total Rebounds
- opptAbbr: Opponent Name
- opptConf: Opponent Conference
- opptDiv: Opponent Division
- opptLoc: Opponent Location
- opptRslt: Opponents Result
- opptDayOff Opponents days off rest


#### Team

In [20]:
#exploring team
team.index
team.describe()
team.columns
columns = list(team) 
columns
featteam = list(team)
team.isnull().any(axis = 1).sum()
team = team.dropna(axis = 0, how='any')
team = team.reset_index(drop=True)

Team is a dataframe with 123 variables and 14758 events, and 4 rows with missing values. Variable are:
- gmDate: Game date
- gmTime: Starting game time
- seasTyp: playoff or regular season
- offLNm1: First official last name
- offFNm1: First official first name
- offLNm2: Second official last name
- offFNm2: Second official first name
- offLNm3: Second offical last name
- offFNm3: Second official first name
- teamAbbr: team name
- teamConf: team conference
- teamDiv: team division
- teamLoc: team location
- teamRslt: team result
- teamMin: teams minutes
- teamDayOff: teams days of rest
- teamPTS: team points
- teamAST: team assists
- teamTO: team turnovers
- teamSTL: teams steals
- teamBLK: team blocks
- teamPF: team personal fouls
- teamFGA: team field goal attempter
- teamFGM: team field goal made
- teamFG%: team field goal percentage
- team2PA: team 2 points attempted
- team2PM: team 2 points made
- team2P%: team 2 points percentage
- team3PA: team three points attempted
- team3PM: team three points made
- team3P%: team three point percentage
- teamFTA: team free throws attempted
- teamFTM: team free throws made
- teamFT%: team free throw percentage
- teamORB: team offensive rebounds
- teamDRB: team defensive rebounds
- teamTRB: team total rebounds
- teamPTS1: team points in the first quarter
- teamPTS2: team points in the second quarter
- teamPTS3: team points in the third quarter
- teamPTS4: team points in the fourth quarter
- teamPTS5: team points in the first overtime
- teamPTS6: team points in the second overtime
- teamPTS7: team points in third overtime
- teamPTS8: team points in fourth overtime
- teamTREB%: team total rebound percentage
- teamASST%: team assist percentage
- teamTS%: team true shooting percentage
- teamEFG%: team effective field goal percentage
- teamOREB%: team  offensive rebound percentage
- teamDREB%: team defensive rebound percentage
- teamTO%: team turnover percentage
- teamSTL%: team steal percentage
- teamBLK%: team block percentage
- teamBLKR: team block rate
- teamPPS: team point per shot
- teamFIC: team floor impact counter
- teamFIC40: 
- teamOrtg: team offensive rating
- teamDrtg: team defensive rating
- teamEDiff:
- teamPlay%:
- teamAR:
- teamAST/TO: team assists per turnover ratio
- teamSTL/TO: team steal per turnover ratio
- opptAbbr: opponent name
- opptConf: opponent conference
- opptDiv: opponent division
- opptLoc: wether the opponent is home or away
- opptRslt: opponent result
- opptMin: opponents minutes
- opptDayOff: opponents day off before game
- opptPTS: opponents points
- opptAST: opponents assists
- opptTO: opponents turnovers
- opptSTL: opponents steals
- opptBLK: opponents blocks
- opptPF: opponents personal fouls
- opptFGA: opponents field goals attempter
- opptFGM: opponents field goals made
- opptFG%: opponents field goal percentage
- oppt2PA: opponents 2 pointers attempter
- oppt2PM: opponents 2 pointers made
- oppt2P%: opponents 2 point percentage
- oppt3PA: opponents 3 pointers attempted
- oppt3PM: opponents 3 pointers made
- oppt3P%: opponents 3 point percentage
- opptFTA: opponents free throw attempted
- opptFTM: opponents free throw made
- opptFT%: opponents free throw percentage
- opptORB: opponents ofensive rebounds
- opptDRB: opopnents defensive rebounds
- opptTRB: opponents total rebounds
- opptPTS1: opponents points in the first quarter
- opptPTS2: opponents points in the second quarter
- opptPTS3: opponents points in the third quarter
- opptPTS4: opponents points in the fourth quarter
- opptPTS5: oppoentns points in first overtime
- opptPTS6: opponents points in the second overtime
- opptPTS7: opponents points in the third overtime
- opptPTS8: opponents points in the fourth overtime
- opptTREB%: opponent total rebound percentage
- opptASST%: opponent assist percentage
- opptTS%: opponent true shooting percentage
- opptEFG%: opponent effective field goal percentage
- opptOREB%: opponent offensive rebound percentage
- opptDREB%: opponent defensive rebound percentage
- opptTO%: opponent turnover percentage
- opptSTL%: opponent steal percentage
- opptBLK%: opponent block percentage
- opptBLKR: opponent block rate
- opptPPS: opponent point per shot
- opptFIC: opponent floor impact counter
- opptFIC40:
- opptOrtg- opponent offensive rating
- opptDrtg- opponent defensive rating
- opptEDiff:
- opptPlay%:
- opptAR:
- opptAST/TO- opponent assist per turnover ratio
- opptSTL/TO - opponent steal per turnover ratio
- poss- possesions per game
- pace- pace of the game


### Data preparation
In the previous section we already removed the NA values, but other cleaning processes can be applied.


We started by comparing the features available on each data set.

In [21]:
ds = [featstandings, featboxscore, featteam,featplayer]
names = ['featstandings', 'featboxscore', 'featteam','featplayer']
ds[0]
i=0
for i in range (0,3):
    for y in range (i+1,4):
        x = len(set(ds[i]) & set(ds[y]))
        print(names[i],'(',len(ds[i]), ') and', names[y],'(',len(ds[y]), ') - ', x )

len(set(featboxscore) or set(featteam) or set(featplayer) or set(featstandings))
#set(featboxscore).difference(featteam)

featstandings ( 39 ) and featboxscore ( 119 ) -  1
featstandings ( 39 ) and featteam ( 123 ) -  1
featstandings ( 39 ) and featplayer ( 51 ) -  1
featboxscore ( 119 ) and featteam ( 123 ) -  117
featboxscore ( 119 ) and featplayer ( 51 ) -  15
featteam ( 123 ) and featplayer ( 51 ) -  21


119

With this we can see that boxscore and team share almost all features, therefore we decided to exclude from this analysis the boxscore dataframe and work with team dataframe only instead.

Then we verified if any of the remaining dataframe had features that were not informant, i.e., had one single value.

In [22]:
#verificar se ha variaveis q nao tem valores diferentes
datasets = [standings,team,player]
for y in datasets:
    x = y.describe(include = 'all')
    n = list(x)
    for i in n:
        if x.loc['unique',i] == 1 or x.loc['std',i] == 0:
            print(i)
 

seasTyp
seasTyp


We notice then that seasTyp has only one value and thus removed this feature from the standings and players dataframes.

In [23]:
#coluna seasTyp so tem um elemento
team = team.drop('seasTyp', axis = 1)
player = player.drop('seasTyp', axis = 1)

In [24]:
#unlike the other dataframes, the game date column had a different name - pq e da jornada? sera que faz sentido mudar o noem?
standings.rename(columns={'stDate': 'gmDate'}, inplace=True)

In [12]:
subsetT = team.copy()
cols_ = ['gmDate','offNm1','offNm2','offNm3','teamAbbr']
subsetT = subsetT[cols_]
subsetT

KeyError: "['offNm3', 'offNm2', 'offNm1'] not in index"

On the Team dataframe the officials' first and last names are separated into two columns, which is not relevant. We decided to joing them in one column, and rename that column appropiately.

In [25]:
#colocar os nomes dos arbitros numa so coluna
first = ['offFNm1','offFNm2','offFNm3']
last = ['offLNm1','offLNm2','offLNm3']
for y in range(0,3):
    for i in range (0,len(team)):
        team.loc[i,first[y]] = team.loc[i,first[y]] + ' ' + team.loc[i,last[y]]
    #eliminar a coluna redundante 
    team = team.drop(last[y], axis=1)
    
team = team.rename(columns={"offFNm1": "offNm1","offFNm2": "offNm2","offFNm3": "offNm3"})

It then became clear that there were replicated values on the Team dataframe, two rows for the same game, one started by describing the home team and the other one the away team, and halfway through the columns it switch back, i.e., the team on line i is the same as the team on the opponent in the line i+1, and vice versa. We decided to eliminate the away rows, and delete the teamLoc feature as it became not informative. 

In [ ]:
subsetDOUBLE = team.copy()
cols_ = ['gmDate','teamAbbr','opptAbbr']
subsetDOUBLE = subsetDOUBLE[cols_]
subsetDOUBLE

In [26]:
#has duplicated values (home and away), remove 
#select rows to delete
drop = []
for i in range(0,len(team)):
    if team.loc[i,'teamLoc'] == 'Away':
         drop.append(i)
#delete rows
team = team.drop(team.index[drop])
#check if is unique
team['teamLoc'].describe()
#drop columm because it has no info (one value only)
team = team.drop('teamLoc', axis = 1)

We decided to merge the data sets

In [27]:
standings = standings.drop(columns=['opptGmPlay','opptGmWon','opptOpptGmPlay','opptOpptGmWon'])

O standings da estatisticas A

In [28]:
featstandings = list(standings)
featteam = list(team)
x = list(set(featstandings) & set(featteam))
fulldata = pd.merge(team,standings, on=x, how='inner')
fulldata = pd.merge(fulldata,standings,left_on=  ['gmDate', 'opptAbbr'], right_on= ['gmDate', 'teamAbbr'], how ='inner')
fulldata

gmDate gmTime             offNm1          offNm2            offNm3  \
0     2012-10-30  19:00      Tony Brothers   Michael Smith  Haywoode Workman   
1     2012-10-30  20:00    Monty McCutchen     Sean Wright   Kane Fitzgerald   
2     2012-10-30  22:30       Scott Foster  Gary Zielinski        Eric Dalen   
3     2012-10-31  19:00  Bennett Salvatore      Tony Brown        Eric Lewis   
4     2012-10-31  19:00       Bill Spooner   David Guthrie      Mark Lindsay   
...          ...    ...                ...             ...               ...   
7344  2018-04-11  08:00        Sean Wright   Dedric Taylor       Sean Corbin   
7345  2018-04-11  08:00     James Williams       Karl Lane         Ken Mauer   
7346  2018-04-11  10:30      Ron Garretson     Rodney Mott  Derek Richardson   
7347  2018-04-11  10:30       Kevin Cutler    Bill Kennedy        Eric Lewis   
7348  2018-04-11  10:30         Josh Tiven        J.T. Orr      Scott Foster   

     teamAbbr_x teamConf    teamDiv teamRslt  teamMin  teamDayOff  teamPTS  \
0           CLE     East    Central      Win      240           0       94   
1           MIA     East  Southeast      Win      240           0      120   
2           LAL     West    Pacific     Loss      240           0       91   
3           PHI     East   Atlantic      Win      240           0       84   
4           TOR     East   Atlantic     Loss      240           0       88   
...         ...      ...        ...      ...      ...         ...      ...   
7344        ORL     East  Southeast      Win      239           2      101   
7345        PHI     East   Atlantic      Win      240           1      130   
7346        LAC     West    Pacific     Loss      240           2      100   
7347        POR     West  Northwest      Win      240           2      102   
7348        SAC     West    Pacific      Win      240           2       96   

      teamAST  teamTO  teamSTL  teamBLK  teamPF  teamFGA  teamFGM  teamFG%  \
0          22      21        7        5      21       79       36   0.4557   
1          25       8        8        5      20       79       43   0.5443   
2          24      14        6        5      21       77       38   0.4935   
3          18      16       13       11      14       85       30   0.3529   
4          18      10       12        8      18       91       33   0.3626   
...       ...     ...      ...      ...     ...      ...      ...      ...   
7344       20      16        6        7      27       77       33   0.4286   
7345       35      12       15        7      17      105       52   0.4952   
7346       27      17        8        1      17       83       41   0.4940   
7347       19      10        9        9      22       89       41   0.4607   
7348       22      11        6        3      14       80       38   0.4750   

      team2PA  team2PM  team2P%  team3PA  team3PM  team3P%  teamFTA  teamFTM  \
0          59       29   0.4915       20        7   0.3500       22       15   
1          63       35   0.5556       16        8   0.5000       32       26   
2          64       35   0.5469       13        3   0.2308       31       12   
3          60       23   0.3833       25        7   0.2800       21       17   
4          74       27   0.3649       17        6   0.3529       19       16   
...       ...      ...      ...      ...      ...      ...      ...      ...   
7344       45       20   0.4444       32       13   0.4063       27       22   
7345       69       37   0.5362       36       15   0.4167       15       11   
7346       65       35   0.5385       18        6   0.3333       28       12   
7347       65       32   0.4923       24        9   0.3750       16       11   
7348       54       31   0.5741       26        7   0.2692       20       13   

      teamFT%  teamORB  teamDRB  teamTRB  teamPTS1  teamPTS2  teamPTS3  \
0      0.6818       18       36       54        31        19        24   
1      0.8125        5       31       36        31        31      

In [29]:
fulldata = fulldata.rename(columns={"teamAbbr_x": "HomeTeam"})

In [30]:
import re
x = list(fulldata)
for i in range(0,len(x)):
    if '_x' in x[i]:
        y = re.sub('_x', '', x[i])
        y = 'team' + y
        x[i] = y
    if '_y' in x[i]:
        y = re.sub('_y', '', x[i])
        y = 'oppt' + y
        x[i] = y
fulldata.columns = x

O problema do merge é que a data do standings é a data do fim de semana, e no team é a data do jogo em si. e é a posiçao de cada equipa.



One of the identified issues is that date and time are not in the correct format. 


In [31]:
#extract a feature with both that and time in the correct format 
for i in range(0,len(fulldata['gmDate'])):
    fulldata.loc[i,'gmDateTime'] = datetime.strptime(fulldata.loc[i,'gmDate'], "%Y-%m-%d")
    x = fulldata.loc[i,'gmTime'].split(':')
    h = int(x[0])
    m = int(x[1])
    fulldata.loc[i,'gmDateTime'] =  fulldata.loc[i,'gmDateTime'].replace(hour=h, minute = m)
fulldata = fulldata.drop('gmDate', axis = 1)
fulldata = fulldata.drop('gmTime', axis = 1)

### Data visualization

In [ ]:
#para verificar q nao ha nomes "repetidos"
asd = list(fulldata.columns)
asd = sorted(asd)
len(asd)
for i in asd:
    x = fulldata[i].describe()
    if fulldata[i].dtype.name == 'object':
        if x['unique'] <= 10:
            print(i)
            print(x['unique'])
    elif fulldata[i].dtype.name == 'float64':
        if x['std'] == 0:
            print(i)
            print('std = 0')

We should drop opptLoc because it is the same value for all events.
we should drop opptRslt because its the oposite of our target (teamRslt)


In [33]:
fulldata

offNm1          offNm2            offNm3 HomeTeam teamConf  \
0         Tony Brothers   Michael Smith  Haywoode Workman      CLE     East   
1       Monty McCutchen     Sean Wright   Kane Fitzgerald      MIA     East   
2          Scott Foster  Gary Zielinski        Eric Dalen      LAL     West   
3     Bennett Salvatore      Tony Brown        Eric Lewis      PHI     East   
4          Bill Spooner   David Guthrie      Mark Lindsay      TOR     East   
...                 ...             ...               ...      ...      ...   
7344        Sean Wright   Dedric Taylor       Sean Corbin      ORL     East   
7345     James Williams       Karl Lane         Ken Mauer      PHI     East   
7346      Ron Garretson     Rodney Mott  Derek Richardson      LAC     West   
7347       Kevin Cutler    Bill Kennedy        Eric Lewis      POR     West   
7348         Josh Tiven        J.T. Orr      Scott Foster      SAC     West   

        teamDiv teamRslt  teamMin  teamDayOff  teamPTS  teamAST  teamTO  \
0       Central      Win      240           0       94       22      21   
1     Southeast      Win      240           0      120       25       8   
2       Pacific     Loss      240           0       91       24      14   
3      Atlantic      Win      240           0       84       18      16   
4      Atlantic     Loss      240           0       88       18      10   
...         ...      ...      ...         ...      ...      ...     ...   
7344  Southeast      Win      239           2      101       20      16   
7345   Atlantic      Win      240           1      130       35      12   
7346    Pacific     Loss      240           2      100       27      17   
7347  Northwest      Win      240           2      102       19      10   
7348    Pacific      Win      240           2       96       22      11   

      teamSTL  teamBLK  teamPF  teamFGA  teamFGM  teamFG%  team2PA  team2PM  \
0           7        5      21       79       36   0.4557       59       29   
1           8        5      20       79       43   0.5443       63       35   
2           6        5      21       77       38   0.4935       64       35   
3          13       11      14       85       30   0.3529       60       23   
4          12        8      18       91       33   0.3626       74       27   
...       ...      ...     ...      ...      ...      ...      ...      ...   
7344        6        7      27       77       33   0.4286       45       20   
7345       15        7      17      105       52   0.4952       69       37   
7346        8        1      17       83       41   0.4940       65       35   
7347        9        9      22       89       41   0.4607       65       32   
7348        6        3      14       80       38   0.4750       54       31   

      team2P%  team3PA  team3PM  team3P%  teamFTA  teamFTM  teamFT%  teamORB  \
0      0.4915       20        7   0.3500       22       15   0.6818       18   
1      0.5556       16        8   0.5000       32       26   0.8125        5   
2      0.5469       13        3   0.2308       31       12   0.3871       15   
3      0.3833       25        7   0.2800       21       17   0.8095       14   
4      0.3649       17        6   0.3529       19       16   0.8421       15   
...       ...      ...      ...      ...      ...      ...      ...      ...   
7344   0.4444       32       13   0.4063       27       22   0.8148        6   
7345   0.5362       36       15   0.4167       15       11   0.7333       17   
7346   0.5385       18        6   0.3333       28       12   0.4286        8   
7347   0.4923       24        9   0.3750       16       11   0.6875        7   
7348   0.5741       26        7   0.2692       20       13   0.6500        6   

      teamDRB  teamTRB  teamPTS1  teamPTS2  teamPTS3  teamPTS4  teamPTS5  \
0          36       54        31        19        24        20         0   
1          31       36        31        31        31        27         0   
2          31       46       

In [39]:
fdfd = lenght.copy()
fdfd = fdfd.sort_values(['HomeTeam', 'gmDateTime'], ascending=[True, True])

cols_ = ['gmDateTime','HomeTeam','teamRslt']
fdfd = fdfd[cols_]
fdfd

gmDateTime HomeTeam teamRslt
17   2012-11-02 19:30:00      ATL     Loss
38   2012-11-04 19:00:00      ATL      Win
53   2012-11-07 19:30:00      ATL      Win
69   2012-11-09 19:30:00      ATL     Loss
89   2012-11-11 15:30:00      ATL     Loss
...                  ...      ...      ...
7279 2018-04-03 08:00:00      WAS     Loss
7294 2018-04-05 08:00:00      WAS     Loss
7302 2018-04-06 07:00:00      WAS     Loss
7333 2018-04-10 08:00:00      WAS      Win
7344 2018-04-11 08:00:00      WAS     Loss

[14698 rows x 3 columns]

In [36]:
lenght

HomeTeam  away          gmDateTime             offNm1          offNm2  \
0         CLE   NaN 2012-10-30 19:00:00      Tony Brothers   Michael Smith   
1         MIA   NaN 2012-10-30 20:00:00    Monty McCutchen     Sean Wright   
2         LAL   NaN 2012-10-30 22:30:00       Scott Foster  Gary Zielinski   
3         PHI   NaN 2012-10-31 19:00:00  Bennett Salvatore      Tony Brown   
4         TOR   NaN 2012-10-31 19:00:00       Bill Spooner   David Guthrie   
...       ...   ...                 ...                ...             ...   
7344      WAS  away 2018-04-11 08:00:00        Sean Wright   Dedric Taylor   
7345      MIL  away 2018-04-11 08:00:00     James Williams       Karl Lane   
7346      LAL  away 2018-04-11 10:30:00      Ron Garretson     Rodney Mott   
7347      UTA  away 2018-04-11 10:30:00       Kevin Cutler    Bill Kennedy   
7348      HOU  away 2018-04-11 10:30:00         Josh Tiven        J.T. Orr   

                offNm3      pace      poss  team2P%  team2PA  team2PM  \
0     Haywoode Workman   88.9409   88.9409   0.4915       59       29   
1      Kane Fitzgerald   94.9832   94.9832   0.5556       63       35   
2           Eric Dalen   91.5790   91.5790   0.5469       64       35   
3           Eric Lewis   94.3715   94.3715   0.3833       60       23   
4         Mark Lindsay   91.5641   91.5641   0.3649       74       27   
...                ...       ...       ...      ...      ...      ...   
7344       Sean Corbin   96.4293   96.0275   0.3770       61       23   
7345         Ken Mauer  104.3883  104.3883   0.4615       65       30   
7346  Derek Richardson  101.7513  101.7513   0.5102       49       25   
7347        Eric Lewis   97.6761   97.6761   0.3788       66       25   
7348      Scott Foster   91.9864   91.9864   0.5000       36       18   

      team3P%  team3PA  team3PM   teamAR  teamASST%  teamAST  teamAST/TO  \
0      0.3500       20        7  16.7072    61.1111       22      1.0476   
1      0.5000       16        8  19.8287    58.1395       25      3.1250   
2      0.2308       13        3  18.6567    63.1579       24      1.7143   
3      0.2800       25        7  14.0362    60.0000       18      1.1250   
4      0.3529       17        6  14.1332    54.5455       18      1.8000   
...       ...      ...      ...      ...        ...      ...         ...   
7344   0.2941       34       10  12.5110    51.5152       17      1.4167   
7345   0.2800       25        7  15.8228    59.4595       22      1.1000   
7346   0.4359       39       17  18.2428    59.5238       25      1.4706   
7347   0.3478       23        8  13.5788    54.5455       18      1.2000   
7348   0.2826       46       13   9.7969    35.4839       11      0.7857   

     teamAbbr  teamBLK  teamBLK%  teamBLKR teamConf  teamDRB  teamDREB%  \
0         NaN        5    5.6217    8.4746     East       36    66.6667   
1         NaN        5    5.2641    7.9365     East       31    81.5789   
2         NaN        5    5.4598    7.8125     West       31    77.5000   
3         NaN       11   11.6561   18.3333     East       33    67.3469   
4         NaN        8    8.7370   10.8108     East       27    75.0000   
...       ...      ...       ...       ...      ...      ...        ...   
7344      WAS        1    1.0414    1.6393     East       32    64.0000   
7345      MIL        5    4.7898    7.6923     East       33    76.7442   
7346      LAL        6    5.8967   12.2449     West       39    79.5918   
7347      UTA        7    7.1665   10.6061     West       39    73.5849   
7348      HOU        6    6.5227   16.6667     West       32    80.0000   

      teamDayOff    teamDiv  teamDrtg  teamEDiff  teamEFG%  teamFG%  teamFGA  \
0              0    Central   94.4447    11.2435    0.5000   0.4557       79   
1              0  Southeast  112.6515    13.6866    0.5949   0.5443       79   
2              0    Pacific  108.1034    -8.7356    0.5130   0.4935       77   
3              0   Atlantic   79.4731     9.5368    0.394

In [35]:
fdcs = fulldata.copy()
#sort columns by name to separate team and oppt stats
fdcs = fdcs.reindex(sorted(fdcs.columns), axis=1)
#get index of border columns
opptb = fdcs.columns.get_loc("oppt2P%")
oppte = fdcs.columns.get_loc("opptwpyth16.5")
teamb = fdcs.columns.get_loc("team2P%")
teame = fdcs.columns.get_loc("teamwpyth16.5")
#create opponent and home team subsets
oppt = fdcs.copy()
oppt = oppt.drop(oppt.iloc[:, teamb:teame+1], axis = 1)
oppt = oppt.drop('HomeTeam', axis = 1)
TEAM = fdcs.copy()
TEAM = TEAM.drop(TEAM.iloc[:, opptb:oppte+1],axis=1)
#rename oppt columns to match the team names
import re
x = list(oppt)
for i in range(0,len(x)):
    if 'oppt' in x[i]:
        y = re.sub('oppt', '', x[i])
        y = 'team' + y
        x[i] = y
oppt.columns = x
oppt = oppt.rename(columns={'teamteamAbbr': 'HomeTeam'})
#add a column to away to "save" the status 
oppt['away']='away'
#concat oppt and team
lenght = pd.concat([TEAM,oppt])

D:\Anaconda\lib\site-packages\ipykernel_launcher.py:28: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.



In [ ]:
#sort by team and date
fdsorted = lenght.sort_values(['HomeTeam', 'gmDateTime'], ascending=[True, True])
fdsorted = fdsorted.reset_index() 

#get index range for each team
coord = []
for i in range(1,len(fdsorted)-1):
    if i == 1:
        x = [fdsorted.loc[0,'HomeTeam'],0]
    if fdsorted.loc[i,'HomeTeam'] != fdsorted.loc[i+1,'HomeTeam']:
        x.append(i)
        coord.append(x)
        x = [fdsorted.loc[i+1,'HomeTeam'],i+1]
    if i == (len(fdsorted)-2):
        x.append(i+1)
        coord.append(x)

In [ ]:
#variables to move to previous game
variaveis_subir = ['teamconfLoss','teamconfWin','teamawayLoss','teamawayWin','teamstkType','teamstkTot',
                   'teamgameBack','teamgameLost','teamgamePlay','teamgameWon','teamhomeLoss','teamhomeWin',
                   'teamlastFive','teamlastTen','teamptsAgnst','teamptsAllow','teamptsDiff','teamptsFor',
                   'teamptsScore','teamrank','teamrankOrd']

for var in variaveis_subir:
    for i in coord:
        inicio = i[1]
        fim = i[2]
        for i in range(fim,inicio+1,-1):
            fdsorted.loc[i,var] = fdsorted.loc[i-1,var]
        if var == 'teamstkType' or var == 'teamrankOrd':
            fdsorted.loc[inicio,var] = "None"
        else: 
            fdsorted.loc[inicio,var] = 0 
            
#variables to calculate last 5 game average
variaveis_media = ['team2PA','teamSTL','teamOrtg','teamSTL%','teamTS%', 'teamORB','teamTREB%', 'teamSTL/TO',
                   'teamTO','teamTRB', 'teamPF','team2PM','team2P%','team3P%','team3PA','team3PM','teamAR',
                   'teamASST%','teamAST','teamAST/TO','teamBLK','teamBLK%','teamBLKR','teamDRB','teamDREB%',
                   'teamDayOff','teamDrtg','teamEDiff','teamEFG%','teamFG%','teamFGA','teamFGM','teamFT%',
                   'teamFTA','teamFTM','teamptsAgnst','teamptsAllow','teamptsDiff','teamptsFor','teamptsScore','teamPTS']

for var in variaveis_media:
    for i in coord:
        inicio = i[1]
        fim = i[2]
        media = fdsorted.loc[inicio:fim,var].mean()
  
        for i in range(fim,inicio+4,-1):
            fdsorted.loc[i,var] = fdsorted.loc[i-5:i-1,var].mean()
        for i in range(inicio+4,inicio,-1):
            y = (i-inicio)
            values = list(fdsorted.loc[i-y:i-1,var])
            values.append(media)
            fdsorted.loc[i,var] = (sum(values) / len(values) )
        fdsorted.loc[inicio,var] = media
        

In [ ]:
#get variables we didnt edit
fdfl = list(fdsorted)
edited = variaveis_media+variaveis_subir
save = list(set(fdfl) - set(edited))
#remove those which much stay
save.remove('gmDateTime')
save.remove('teamRslt')
save.remove('index')
save.remove('offNm1')
save.remove('offNm2')
save.remove('offNm3')
save.remove('teamAbbr')
save.remove('teamDiv')
save.remove('away')
save.remove('HomeTeam')
#drop remaining
fdsorted= fdsorted.drop(columns=save)
#create two copies : one for oppt and one for home team
completeoppt = fdsorted.copy()
completeoppt = completeoppt[completeoppt['away'] == 'away']
completeteam = fdsorted.copy()
completeteam = completeteam[completeteam['away'] != 'away']
completeoppt = completeoppt.rename(columns={'HomeTeam': 'AwayTeam'})
#rename oppt columns
import re
x = list(completeoppt)
for i in range(0,len(x)):
    if 'team' in x[i]:
        y = re.sub('team', '', x[i])
        y = 'oppt' + y
        x[i] = y
completeoppt.columns = x

#drop signal columns
completeoppt = completeoppt.drop('away', axis = 1)
completeteam = completeteam.drop('away', axis = 1)

#get common variables to merge on
opptlist = list(completeoppt)
teamlist = list(completeteam)
x = list(set(opptlist) & set(teamlist))
#merge into a dataframe with one single game per row, with both home team and oppt team stats
finaldf= pd.merge(completeteam,completeoppt, on=x, how='inner')
#drop more redudant variables
finaldf = finaldf.drop('teamAbbr', axis = 1)
finaldf = finaldf.drop('opptAbbr', axis = 1)

finaldf = finaldf.drop('gmDateTime', axis = 1)
#reset original index
finaldf = finaldf.sort_values('index', ascending=True)
finaldf = finaldf.set_index('index')


## Feature engineering
In order to build prediction models we cant have variables regarding statistics of the game we are trying to perdict. That being said, we had to perform feature engineering.


## Modeling 

In [ ]:
#normalization ?

### Get dummy variables

In [ ]:
#remover opptRslt pw é o oposto do target
modeldata = finaldf.copy()
modeldata = modeldata.drop('opptRslt', axis = 1)
#get features and target datasets
dummyX = modeldata.copy()
dummyX = dummyX.drop('teamRslt', axis=1)
dummyY = modeldata.teamRslt
#get categorical variables
cols = dummyX.columns
num_cols = dummyX._get_numeric_data().columns
cat_cols = list(set(cols) - set(num_cols))
#get dummies of cat variables and of target
dummyX = pd.get_dummies(dummyX, columns=cat_cols)
dummy = pd.get_dummies(dummyY)
dummyY = pd.concat([dummyY, dummy], axis=1)
dummyY = dummy.Win
#split data set into train and test
dummyX_train = dummyX.loc[0:int((len(dummyX)*0.7)),:]
dummyX_test = dummyX.loc[(int(len(dummyX)*0.7)):len(dummyX),:]
dummyy_train = dummyY.loc[0:int((len(dummyY)*0.7))]
dummyy_test = dummyY.loc[(int(len(dummyY)*0.7)):]

In [ ]:
models = ['Logistic Regression','K-Nearest Neighbours','C-Support Vector','Linear Support Vector','Gaussian Naive Bayes','Multi-layer Perceptron' ,'Decision Tree','Gradient Boosting','Random Forest']
best_accuracy = [None] * 9
scaled = ['False']*9
modelcomparison = pd.DataFrame(list(zip(models,best_accuracy,scaled)), columns =['Model','Accuracy','Scaled']) 
modelcomparison

## Gaussian Naive Bayes

In [13]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
gnb = GaussianNB().fit(dummyX_train, dummyy_train)
gnbpreds = gnb.predict(dummyX_test)
modelcomparison.iloc[4,1] = accuracy_score(dummyy_test, gnbpreds)


NameError: name 'dummyX_train' is not defined

## Multi-layer Perceptron 

In [ ]:
from sklearn.neural_network import MLPClassifier
#dictionary to store results
MLPCopt = {}
#Lists of parameters to test
act = ['identity', 'logistic', 'tanh', 'relu']
sol = ['lbfgs', 'sgd', 'adam']
lr = ['constant', 'invscaling', 'adaptive']
#iterate over each possible combination
for activator in act:
    for solver in sol:
        if solver == 'sgd' or 'adam':
            shuffle = False
        if solver == 'sgd':
            for rate in lr:
                #build model
                MLPC = MLPClassifier(solver=solver, activation=activator, learning_rate=rate,random_state=0).fit(dummyX_train, dummyy_train)
                #make predictions
                MLPCpreds = MLPC.predict(dummyX_test)
                comb = activator + solver + rate
                #save accuracy result
                MLPCopt[comb] = accuracy_score(dummyy_test, MLPCpreds)
        else: 
            #build model
            MLPC = MLPClassifier(solver=solver, activation=activator,random_state=0).fit(dummyX_train, dummyy_train)
            #make predictions
            MLPCpreds = MLPC.predict(dummyX_test)
            comb = activator + solver
            #save accuracy result
            MLPCopt[comb] = accuracy_score(dummyy_test, MLPCpreds)

#get default parameter model
MLPC = MLPClassifier(random_state=0).fit(dummyX_train, dummyy_train)
MLPCpreds = MLPC.predict(dummyX_test)
MLPCopt['default'] = accuracy_score(dummyy_test, MLPCpreds)
#append best accuracy score to the results of all models
modelcomparison.iloc[5,1] = MLPCopt[max(MLPCopt, key=MLPCopt.get)]

In [ ]:
MLPCdf = pd.DataFrame(list(MLPCopt.items()), columns=['Model_Variant', 'Accuracy'])
MLPCdf = pd.DataFrame(data=list(MLPCdf['Accuracy']),columns=['Accuracy'],index=MLPCdf['Model_Variant'])
MLPCdf = MLPCdf.drop('default',axis=0)
fig,ax= plt.subplots()

MLPCdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(18,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.4,0.7)
# use axhline
ax.axhline(MLPCopt['default'],color='firebrick')
fig.savefig('MLPC1.png', bbox_inches='tight')

## K-Nearest Neihbors

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
#dictionary to store results
KNNopt = {}
MLPCopt = {}
#Lists of parameters to test
nnr = [1,2,3,5,10,20]
wgh = ['uniform','distance']
alg = ['auto', 'ball_tree', 'kd_tree', 'brute']
pi = [1,2]
#iterate over each possible combination
for n in nnr:
    for w in wgh:
        for a in alg:
            for p in pi:
                #build model
                knn = KNeighborsClassifier(n_neighbors=n,weights=w,algorithm=a,p=p).fit(dummyX_train, dummyy_train)                
                #make predictions
                knnpred = knn.predict(dummyX_test)
                comb = str(n)+w+a+str(p)
                #save accuracy result
                KNNopt[comb]=accuracy_score(dummyy_test, knnpred)

#get default parameter model
knn = KNeighborsClassifier().fit(dummyX_train, dummyy_train)
knnpred = knn.predict(dummyX_test)
KNNopt['default']=accuracy_score(dummyy_test, knnpred)
#append best accuracy score to the results of all models
modelcomparison.iloc[1,1]=KNNopt[max(KNNopt, key=KNNopt.get)]

In [ ]:
knndf = pd.DataFrame(list(KNNopt.items()), columns=['Model_Variant', 'Accuracy'])
knndf = pd.DataFrame(data=list(knndf['Accuracy']),columns=['Accuracy'],index=knndf['Model_Variant'])
knndf = knndf.drop('default',axis=0)
fig,ax= plt.subplots()

knndf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(20,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.50,0.64)
# use axhline
ax.axhline(KNNopt['default'],color='firebrick')
fig.savefig('KNN1.png', bbox_inches='tight')

In [ ]:
improvement = knndf['Accuracy']>=KNNopt['default']
knndf = knndf[improvement]
fig,ax= plt.subplots()

knndf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(14,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.58,0.64)
# use axhline
ax.axhline(KNNopt['default'],color='firebrick')
fig.savefig('KNN2.png', bbox_inches='tight')

In [ ]:
best = knndf['Accuracy']>=0.622
knndf[best]

## Logistic Regression

In [ ]:
from sklearn import linear_model
#dictionary to store results
LRopt = {}
#Lists of parameters to test
sol = ['newton-cg', 'lbfgs', 'liblinear', 'sag','saga']
pen = ['l1','l2','none']
#iterate over each possible combination
for solver in sol:
    if solver == 'newton-cg' or solver == 'sag' or solver == 'lbfgs':
        #build model
        regr = linear_model.LogisticRegression(solver=solver, penalty='l2',random_state=0).fit(dummyX_train, dummyy_train)
        #make predictions
        regr_pred = regr.predict(dummyX_test)
        comb = solver + 'l2'
        #save accuracy result
        LRopt[comb] = accuracy_score(dummyy_test,regr_pred)
        
        #build model
        regr = linear_model.LogisticRegression(solver=solver,penalty='none',random_state=0).fit(dummyX_train, dummyy_train)
        #make predictions
        regr_pred = regr.predict(dummyX_test)
        comb = solver + 'none'
        #save accuracy result
        LRopt[comb] = accuracy_score(dummyy_test,regr_pred)
        
    if solver == 'saga':
        for penalty in pen:
            #build model
            regr = linear_model.LogisticRegression(solver=solver, penalty=penalty,random_state=0).fit(dummyX_train, dummyy_train)
            #make predictions
            regr_pred = regr.predict(dummyX_test)
            comb = solver + penalty
            #save accuracy result
            LRopt[comb] = accuracy_score(dummyy_test,regr_pred)
            
    if solver == 'liblinear':
        for penalty in pen[0:1]:
            #build model
            regr = linear_model.LogisticRegression(solver=solver, penalty=penalty,random_state=0).fit(dummyX_train, dummyy_train)
            #make predictions
            regr_pred = regr.predict(dummyX_test)
            comb = solver + penalty
            #save accuracy result
            LRopt[comb] = accuracy_score(dummyy_test,regr_pred)

#get default parameter model
regr = linear_model.LogisticRegression(random_state=0).fit(dummyX_train, dummyy_train)
regr_pred = regr.predict(dummyX_test)
LRopt['default'] = accuracy_score(dummyy_test,regr_pred)
            
#append best accuracy score to the results of all models
modelcomparison.iloc[0,1]=LRopt[max(LRopt, key=LRopt.get)] 

In [ ]:
lrdf = pd.DataFrame(list(LRopt.items()), columns=['Model_Variant', 'Accuracy'])
lrdf = pd.DataFrame(data=list(lrdf['Accuracy']),columns=['Accuracy'],index=lrdf['Model_Variant'])
lrdf = lrdf.drop('default',axis=0)
fig,ax= plt.subplots()

lrdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(10,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.54,0.67)
# use axhline
ax.axhline(LRopt['default'],color='firebrick')
fig.savefig('LR1.png', bbox_inches='tight')


## C-Support Vector Classification

In [ ]:
from sklearn.svm import SVC
SVCmodel = SVC(random_state=0).fit(dummyX_train, dummyy_train) 
SVCpreds = SVCmodel.predict(dummyX_test)
modelcomparison.iloc[2,1]= accuracy_score(dummyy_test, SVCpreds)

In [ ]:
#SVCopt = {}
#kernel =['linear','poly','rbf','sigmoid','precomputed']
#gamma = ['scale','auto'] 
#for k in kernel :
#    if k == 'rbf' or k == 'poly' or k == 'sigmoid':
#        for g in gamma:            
#            SVCmodel = SVC(gamma=g,kernel=k).fit(dummyX_train, dummyy_train) 
#            SVCpreds = SVCmodel.predict(dummyX_test)
#            comb = k+g
#            SVCopt[comb] = accuracy_score(dummyy_test, SVCpreds)
#    else:
#        SVCmodel = SVC(kernel=k).fit(dummyX_train, dummyy_train) 
#        SVCpreds = SVCmodel.predict(dummyX_test)
#        comb = k
#        SVCopt[comb] = accuracy_score(dummyy_test, SVCpreds)


## Linear Support Vector Classification

In [ ]:
from sklearn.svm import LinearSVC
LSVC = LinearSVC(random_state=0).fit(dummyX_train, dummyy_train)
LSVCpreds = LSVC.predict(dummyX_test)
modelcomparison.iloc[3,1]= accuracy_score(dummyy_test, LSVCpreds)



## Gradient Boosting Classifier

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
gtb = GradientBoostingClassifier(random_state=0).fit(dummyX_train, dummyy_train)
gtbpreds = gtb.predict(dummyX_test)
modelcomparison.iloc[7,1]= accuracy_score(dummyy_test, gtbpreds)


In [ ]:
#loss = ['deviance','exponential']
#crit =['friedman_mse','mse','mae']
#max_f = ['sqrt','log2', 'none']
#l_rate = [0.01,0.1,0.5]
#max_d = [2,4,8,16,33]
#n_est = [25,50,100,200,400]
#GTBopt = {}
#for c in crit:
#    for n in n_est:
#        for md in max_d:
#            for l in l_rate:
#                for ss in loss:
#                    for mf in max_f:
#                        if mf == 'none':
#                            gtb = GradientBoostingClassifier(criterion=c,n_estimators=n,learning_rate=l,loss=ss, max_depth=md).fit(dummyX_train,dummyy_train)
#                        else:
#                            gtb = GradientBoostingClassifier(criterion=c,n_estimators=n,max_features=mf,learning_rate=l,loss=ss, max_depth=md).fit(dummyX_train,dummyy_train)
#                        gtbpreds = gtb.predict(dummyX_test)
#                        comb = c+str(n)+str(l)+ss+mf+str(md)
#                        GTBopt[comb] = metrics.accuracy_score(dummyy_test, gtbpreds)
#
#GTBopt
#max(GTBopt, key=GTBopt.get)

## Decision Tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier
#dictionary to store the results
DTCopt = {}
#list of parameters to test
crit = ['gini','entropy']
spl = ['best','random']
max_f = ['sqrt','log2', 'none']
max_d = [2,4,8,16,33]
#iterate over each possible combination 
for c in crit:
    for s in spl:
        for md in max_d:
            for mf in max_f:
                if mf == 'none':
                    #build model
                    DTC = DecisionTreeClassifier(criterion=c, splitter=s, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                else:
                    #build model
                    DTC = DecisionTreeClassifier(criterion=c, splitter=s,max_features=mf, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                #make predictions
                DTCpreds = DTC.predict(dummyX_test)
                comb = c+s+mf+str(md)
                #save accuracy results
                DTCopt[comb] = accuracy_score(dummyy_test, DTCpreds)

#get default parameter model
DTC = DecisionTreeClassifier(random_state=0).fit(dummyX_train,dummyy_train)
DTCpreds = DTC.predict(dummyX_test)
DTCopt['default'] = accuracy_score(dummyy_test, DTCpreds)
#append best accuracy score to the results of all models
modelcomparison.iloc[6,1]=DTCopt[max(DTCopt, key=DTCopt.get)] 

In [ ]:
dtcdf = pd.DataFrame(list(DTCopt.items()), columns=['Model_Variant', 'Accuracy'])
dtcdf = pd.DataFrame(data=list(dtcdf['Accuracy']),columns=['Accuracy'],index=dtcdf['Model_Variant'])
dtcdf = dtcdf.drop('default',axis=0)
fig,ax= plt.subplots()

dtcdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(10,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.50,0.66)
# use axhline
ax.axhline(DTCopt['default'],color='firebrick')
fig.savefig('DTC1.png', bbox_inches='tight')

In [ ]:
dtcdf = dtcdf.nlargest(5,columns='Accuracy')
fig,ax= plt.subplots()

dtcdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(5,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.62,0.65)
fig.savefig('DTC2.png', bbox_inches='tight')


## Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
#dictionary to store the results 
RFCopt = {}
#lists of parameters to test
n_est = [25,50,100,200,400]
boot = [True,False]
crit = ['gini','entropy']
max_f = ['sqrt','log2', 'none']
max_d = [2,4,8,16,33]
#iterate over each possible combination
for c in crit:
    for n in n_est:
        for md in max_d:
            for b in boot:
                for mf in max_f:
                    if mf == 'none':
                        #build model
                        RFC = RandomForestClassifier(criterion=c,n_estimators=n,bootstrap=b, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                    else:
                        #build model
                        RFC = RandomForestClassifier(criterion=c,n_estimators=n,bootstrap=b,max_features=mf, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                    #make predictions
                    RFCpred = RFC.predict(dummyX_test)
                    comb = c+str(n)+str(b)+mf+str(md)
                    #save accuracy result
                    RFCopt[comb] = accuracy_score(dummyy_test, RFCpred )

#get default parameter model
RFC = RandomForestClassifier(random_state=0).fit(dummyX_train,dummyy_train)
RFCpred = RFC.predict(dummyX_test)
RFCopt['default'] = accuracy_score(dummyy_test, RFCpred )                    
#append best accuracy score to the results of all models
modelcomparison.iloc[8,1]=RFCopt[max(RFCopt, key=RFCopt.get)] 

In [ ]:
rfcdf = pd.DataFrame(list(RFCopt.items()), columns=['Model_Variant', 'Accuracy'])
rfcdf = pd.DataFrame(data=list(rfcdf['Accuracy']),columns=['Accuracy'],index=rfcdf['Model_Variant'])
rfcdf = rfcdf.drop('default',axis=0)
rfcdf = rfcdf.nlargest(15,columns='Accuracy')
#testou 300 modelos/238 melhoraram
#improvement = rfcdf['Accuracy']>=RFCopt['default']
#rfcdf = rfcdf[improvement]
fig,ax= plt.subplots()

rfcdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(10,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.61,0.67)
# use axhline
ax.axhline(RFCopt['default'],color='firebrick')
fig.savefig('RF1.png', bbox_inches='tight')

In [ ]:
rfcdf

In [ ]:
modelcomparison.plot(kind='bar',x='Model',y='Accuracy',color='cadetblue').get_legend().remove()
plt.ylabel('Accuracy')
plt.ylim(0.4,0.7)
plt.savefig('compare1.png', bbox_inches='tight')

In [ ]:
import seaborn as sns
num_col = completeteam._get_numeric_data().columns
histdf = completeteam[num_col]
#for i, col in enumerate(histdf.columns):
#    plt.figure(i)
#    sns.distplot(histdf[col])
fig, axes = plt.subplots(nrows=4, ncols=14, figsize=(40,14))
axes = axes.flatten()
for ax, col in zip(axes, histdf.columns):
    sns.distplot(histdf[col], ax=ax)
#plt.show()
#histdf.plot(kind="barh", subplots=True, layout=(4,14), legend=False)
#
plt.tight_layout()  

fig.savefig('histo.png', bbox_inches='tight', dpi=600)
#plt.show()

In [ ]:
axes.flatten()

In [ ]:
num_col = finaldf._get_numeric_data().columns
stdvar = list(num_col)
stdvar.remove('teamawayLoss')
stdvar.remove('teamawayWin')
stdvar.remove('teamconfLoss')
stdvar.remove('teamconfWin')
stdvar.remove('teamgameBack')
stdvar.remove('teamgameLost')
stdvar.remove('teamgameWon')
stdvar.remove('teamgamePlay')
stdvar.remove('teamhomeLoss')
stdvar.remove('teamhomeWin')
stdvar.remove('teamptsAgnst')
stdvar.remove('teamptsFor')
stdvar.remove('teamrank')
stdvar.remove('teamstkTot')
stdvar.remove('opptawayLoss')
stdvar.remove('opptawayWin')
stdvar.remove('opptconfLoss')
stdvar.remove('opptconfWin')
stdvar.remove('opptgameBack')
stdvar.remove('opptgameLost')
stdvar.remove('opptgameWon')
stdvar.remove('opptgamePlay')
stdvar.remove('oppthomeLoss')
stdvar.remove('oppthomeWin')
stdvar.remove('opptptsAgnst')
stdvar.remove('opptptsFor')
stdvar.remove('opptrank')
stdvar.remove('opptstkTot')
normvar = list(set(num_col) - set(stdvar))

In [ ]:
stddf = finaldf[stdvar]
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler() 
stdscaled = scaler.fit_transform(stddf)
stddf = pd.DataFrame(stdscaled, index=stddf.index, columns=stddf.columns)

normdf = finaldf[normvar]
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler() 
normscaled = scaler.fit_transform(normdf)
normdf = pd.DataFrame(normscaled, index=normdf.index, columns=normdf.columns)

#remover opptRslt pw é o oposto do target
scaled_data = finaldf.copy()
scaled_data = scaled_data.drop('opptRslt', axis = 1)
scaled_data = scaled_data.drop(stdvar,axis=1)
scaled_data = scaled_data.drop(normvar,axis=1)
scaled_data = scaled_data.merge(stddf,right_index=True, left_index=True)
scaled_data = scaled_data.merge(normdf,right_index=True, left_index=True)

#get features and target datasets
dummyX = scaled_data.copy()
dummyX = dummyX.drop('teamRslt', axis=1)
dummyY = scaled_data.teamRslt
#get categorical variables
cols = dummyX.columns
num_cols = dummyX._get_numeric_data().columns
cat_cols = list(set(cols) - set(num_cols))
#get dummies of cat variables and of target
dummyX = pd.get_dummies(dummyX, columns=cat_cols)
dummy = pd.get_dummies(dummyY)
dummyY = pd.concat([dummyY, dummy], axis=1)
dummyY = dummy.Win
#split data set into train and test
dummyX_train = dummyX.loc[0:int((len(dummyX)*0.7)),:]
dummyX_test = dummyX.loc[(int(len(dummyX)*0.7)):len(dummyX),:]
dummyy_train = dummyY.loc[0:int((len(dummyY)*0.7))]
dummyy_test = dummyY.loc[(int(len(dummyY)*0.7)):]

In [ ]:
models = ['Logistic Regression','K-Nearest Neighbours','C-Support Vector','Linear Support Vector','Gaussian Naive Bayes','Multi-layer Perceptron' ,'Decision Tree','Gradient Boosting','Random Forest']
best_accuracy = [None] * 9
scaled = ['True']*9
scaledcomparison = pd.DataFrame(list(zip(models,best_accuracy,scaled)), columns =['Model','Accuracy','Scaled']) 
scaledcomparison

In [ ]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
gnb = GaussianNB().fit(dummyX_train, dummyy_train)
gnbpreds = gnb.predict(dummyX_test)
scaledcomparison.iloc[4,1] = accuracy_score(dummyy_test, gnbpreds)


In [ ]:
from sklearn.neural_network import MLPClassifier
scaledMLPCopt = {}
act = ['identity', 'logistic', 'tanh', 'relu']
sol = ['lbfgs', 'sgd', 'adam']
lr = ['constant', 'invscaling', 'adaptive']

for activator in act:
    for solver in sol:
        if solver == 'sgd' or 'adam':
            shuffle = False
        if solver == 'sgd':
            for rate in lr:
                MLPC = MLPClassifier(solver=solver, activation=activator, learning_rate=rate,random_state=0).fit(dummyX_train, dummyy_train)
                MLPCpreds = MLPC.predict(dummyX_test)
                comb = activator + solver + rate
                scaledMLPCopt[comb] = accuracy_score(dummyy_test, MLPCpreds)
        else: 
            MLPC = MLPClassifier(solver=solver, activation=activator,random_state=0).fit(dummyX_train, dummyy_train)
            MLPCpreds = MLPC.predict(dummyX_test)
            comb = activator + solver
            scaledMLPCopt[comb] = accuracy_score(dummyy_test, MLPCpreds)


MLPC = MLPClassifier(random_state=0).fit(dummyX_train, dummyy_train)
MLPCpreds = MLPC.predict(dummyX_test)
scaledMLPCopt['default'] = accuracy_score(dummyy_test, MLPCpreds)
#append best accuracy score to the results df
scaledcomparison.iloc[5,1] = scaledMLPCopt[max(scaledMLPCopt, key=scaledMLPCopt.get)]

In [ ]:
scaledMLPCdf = pd.DataFrame(list(scaledMLPCopt.items()), columns=['Model_Variant', 'Accuracy'])
scaledMLPCdf = pd.DataFrame(data=list(scaledMLPCdf['Accuracy']),columns=['Accuracy'],index=scaledMLPCdf['Model_Variant'])
scaledMLPCdf = scaledMLPCdf.drop('default',axis=0)
fig,ax= plt.subplots()

scaledMLPCdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(20,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.4,0.7)
# use axhline
ax.axhline(scaledMLPCopt['default'],color='firebrick')
fig.savefig('normMLPC1.png', bbox_inches='tight')

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
scaledKNNopt = {}
nnr = [1,2,3,5,10,20]
wgh = ['uniform','distance']
alg = ['auto', 'ball_tree', 'kd_tree', 'brute']
pi = [1,2]
for n in nnr:
    for w in wgh:
        for a in alg:
            for p in pi:
                knn = KNeighborsClassifier(n_neighbors=n,weights=w,algorithm=a,p=p).fit(dummyX_train, dummyy_train)
                knnpred = knn.predict(dummyX_test)
                comb = str(n)+w+a+str(p)
                scaledKNNopt[comb]=accuracy_score(dummyy_test, knnpred)
knn = KNeighborsClassifier().fit(dummyX_train, dummyy_train)
knnpred = knn.predict(dummyX_test)
scaledKNNopt['default']=accuracy_score(dummyy_test, knnpred)
#append best accuracy score to the results df
scaledcomparison.iloc[1,1]=scaledKNNopt[max(scaledKNNopt, key=scaledKNNopt.get)]

In [ ]:
scaledknndf = pd.DataFrame(list(scaledKNNopt.items()), columns=['Model_Variant', 'Accuracy'])
scaledknndf = pd.DataFrame(data=list(scaledknndf['Accuracy']),columns=['Accuracy'],index=scaledknndf['Model_Variant'])
scaledknndf = scaledknndf.drop('default',axis=0)
fig,ax= plt.subplots()

scaledknndf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(20,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.50,0.64)
# use axhline
ax.axhline(scaledKNNopt['default'],color='firebrick')
fig.savefig('normKNN1.png', bbox_inches='tight')

In [ ]:
improvement = scaledknndf['Accuracy']>=scaledKNNopt['default']
scaledknndf = scaledknndf[improvement]
fig,ax= plt.subplots()

scaledknndf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(14,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.58,0.64)
# use axhline
ax.axhline(scaledKNNopt['default'],color='firebrick')
fig.savefig('normKNN2.png', bbox_inches='tight')

In [ ]:
from sklearn import linear_model
scaledLRopt = {}
sol = ['newton-cg', 'lbfgs', 'liblinear', 'sag','saga']
pen = ['l1','l2','none']
for solver in sol:
    if solver == 'newton-cg' or solver == 'sag' or solver == 'lbfgs':
        regr = linear_model.LogisticRegression(solver=solver, penalty='l2',random_state=0).fit(dummyX_train, dummyy_train)
        regr_pred = regr.predict(dummyX_test)
        comb = solver + 'l2'
        scaledLRopt[comb] = accuracy_score(dummyy_test,regr_pred)
        regr = linear_model.LogisticRegression(solver=solver,penalty='none',random_state=0).fit(dummyX_train, dummyy_train)
        regr_pred = regr.predict(dummyX_test)
        comb = solver + 'none'
        scaledLRopt[comb] = accuracy_score(dummyy_test,regr_pred)
        
    if solver == 'saga':
        for penalty in pen:
            regr = linear_model.LogisticRegression(solver=solver, penalty=penalty,random_state=0).fit(dummyX_train, dummyy_train)
            regr_pred = regr.predict(dummyX_test)
            comb = solver + penalty
            scaledLRopt[comb] = accuracy_score(dummyy_test,regr_pred)
    if solver == 'liblinear':
        for penalty in pen[0:1]:
            regr = linear_model.LogisticRegression(solver=solver, penalty=penalty,random_state=0).fit(dummyX_train, dummyy_train)
            regr_pred = regr.predict(dummyX_test)
            comb = solver + penalty
            scaledLRopt[comb] = accuracy_score(dummyy_test,regr_pred)

regr = linear_model.LogisticRegression(random_state=0).fit(dummyX_train, dummyy_train)
regr_pred = regr.predict(dummyX_test)
scaledLRopt['default'] = accuracy_score(dummyy_test,regr_pred)
            
#append best accuracy score to the results df
scaledcomparison.iloc[0,1]=scaledLRopt[max(scaledLRopt, key=scaledLRopt.get)] 

In [ ]:
scaledlrdf = pd.DataFrame(list(scaledLRopt.items()), columns=['Model_Variant', 'Accuracy'])
scaledlrdf = pd.DataFrame(data=list(scaledlrdf['Accuracy']),columns=['Accuracy'],index=scaledlrdf['Model_Variant'])
scaledlrdf = scaledlrdf.drop('default',axis=0)
fig,ax= plt.subplots()

scaledlrdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(10,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.54,0.67)
# use axhline
ax.axhline(scaledLRopt['default'],color='firebrick')
fig.savefig('normLR1.png', bbox_inches='tight')


In [ ]:
from sklearn.svm import SVC
SVCmodel = SVC(random_state=0).fit(dummyX_train, dummyy_train) 
SVCpreds = SVCmodel.predict(dummyX_test)
scaledcomparison.iloc[2,1]= accuracy_score(dummyy_test, SVCpreds)

In [ ]:
from sklearn.svm import LinearSVC
LSVC = LinearSVC(random_state=0).fit(dummyX_train, dummyy_train)
LSVCpreds = LSVC.predict(dummyX_test)
scaledcomparison.iloc[3,1]= accuracy_score(dummyy_test, LSVCpreds)



In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
gtb = GradientBoostingClassifier(random_state=0).fit(dummyX_train, dummyy_train)
gtbpreds = gtb.predict(dummyX_test)
scaledcomparison.iloc[7,1]= accuracy_score(dummyy_test, gtbpreds)


In [ ]:
from sklearn.tree import DecisionTreeClassifier
crit = ['gini','entropy']
spl = ['best','random']
max_f = ['sqrt','log2', 'none']
max_d = [2,4,8,16,33]
scaledDTCopt = {}
for c in crit:
    for s in spl:
        for md in max_d:
            for mf in max_f:
                if mf == 'none':
                    DTC = DecisionTreeClassifier(criterion=c, splitter=s, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                else:
                    DTC = DecisionTreeClassifier(criterion=c, splitter=s,max_features=mf, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                DTCpreds = DTC.predict(dummyX_test)
                comb = c+s+mf+str(md)
                scaledDTCopt[comb] = accuracy_score(dummyy_test, DTCpreds)
                
DTC = DecisionTreeClassifier(random_state=0).fit(dummyX_train,dummyy_train)
DTCpreds = DTC.predict(dummyX_test)
scaledDTCopt['default'] = accuracy_score(dummyy_test, DTCpreds)
scaledcomparison.iloc[6,1]=scaledDTCopt[max(scaledDTCopt, key=scaledDTCopt.get)] 

In [ ]:
scaleddtcdf = pd.DataFrame(list(scaledDTCopt.items()), columns=['Model_Variant', 'Accuracy'])
scaleddtcdf = pd.DataFrame(data=list(scaleddtcdf['Accuracy']),columns=['Accuracy'],index=scaleddtcdf['Model_Variant'])
scaleddtcdf = scaleddtcdf.drop('default',axis=0)
fig,ax= plt.subplots()

scaleddtcdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(10,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.50,0.66)
# use axhline
ax.axhline(scaledDTCopt['default'],color='firebrick')
fig.savefig('normDTC1.png', bbox_inches='tight')

In [ ]:
scaleddtcdf = scaleddtcdf.nlargest(5,columns='Accuracy')
fig,ax= plt.subplots()

scaleddtcdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(5,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.63,0.65)
fig.savefig('normDTC2.png', bbox_inches='tight')

In [ ]:
from sklearn.ensemble import RandomForestClassifier
n_est = [25,50,100,200,400]
boot = [True,False]
crit = ['gini','entropy']
max_f = ['sqrt','log2', 'none']
max_d = [2,4,8,16,33]
scaledRFCopt = {}
for c in crit:
    for n in n_est:
        for md in max_d:
            for b in boot:
                for mf in max_f:
                    if mf == 'none':
                        RFC = RandomForestClassifier(criterion=c,n_estimators=n,bootstrap=b, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                    else:
                        RFC = RandomForestClassifier(criterion=c,n_estimators=n,bootstrap=b,max_features=mf, max_depth=md,random_state=0).fit(dummyX_train,dummyy_train)
                    RFCpred = RFC.predict(dummyX_test)
                    comb = c+str(n)+str(b)+mf+str(md)
                    scaledRFCopt[comb] = accuracy_score(dummyy_test, RFCpred )

RFC = RandomForestClassifier(random_state=0).fit(dummyX_train,dummyy_train)
RFCpred = RFC.predict(dummyX_test)
scaledRFCopt['default'] = accuracy_score(dummyy_test, RFCpred )                    
                    
scaledcomparison.iloc[8,1]=scaledRFCopt[max(scaledRFCopt, key=scaledRFCopt.get)] 

In [ ]:
scaledrfcdf = pd.DataFrame(list(scaledRFCopt.items()), columns=['Model_Variant', 'Accuracy'])
scaledrfcdf = pd.DataFrame(data=list(scaledrfcdf['Accuracy']),columns=['Accuracy'],index=scaledrfcdf['Model_Variant'])
scaledrfcdf = scaledrfcdf.drop('default',axis=0)
scaledrfcdf = scaledrfcdf.nlargest(15,columns='Accuracy')
#testou 300 modelos/238 melhoraram
#improvement = rfcdf['Accuracy']>=RFCopt['default']
#rfcdf = rfcdf[improvement]
fig,ax= plt.subplots()

scaledrfcdf.plot(kind="bar", ax=ax,color='cadetblue',figsize=(10,5)).get_legend().remove()

plt.minorticks_on()
ax.tick_params(axis='x',which='minor',bottom='off')
ax.set_xlabel("Model Variant")
ax.set_ylabel("Accuracy")
ax.set_ylim(0.61,0.67)
# use axhline
ax.axhline(scaledRFCopt['default'],color='firebrick')
fig.savefig('normRF1.png', bbox_inches='tight')

In [ ]:
scaledcomparison.plot(kind='bar',x='Model',y='Accuracy',color='cadetblue').get_legend().remove()
plt.ylabel('Accuracy')
plt.ylim(0.4,0.7)
fig.savefig('normcompare1.png', bbox_inches='tight')

In [ ]:
comparisondf = pd.concat([modelcomparison,scaledcomparison])

In [ ]:
comparisondf = comparisondf.reset_index()
comparisondf

In [ ]:
width = 0.35  # the width of the bars
ind = np.arange(9)
fig, ax = plt.subplots()
rects1 = ax.bar(ind - width/2, modelcomparison['Accuracy'], width, 
                color='cadetblue', label='Unscaled')
rects2 = ax.bar(ind + width/2, scaledcomparison['Accuracy'], width,
                color='firebrick', label='Scaled')
plt.xticks(ind + width / 2, (modelcomparison['Model']),rotation='vertical')
plt.ylim(0.45,0.7)
plt.legend(loc='best')
fig.savefig('unsaceld_vs_scaled.png', bbox_inches='tight')

In [ ]:
Accuracy = []
Accuracy.append(comparisondf.iloc[6,2] )
Accuracy.append(comparisondf.iloc[15,2] )
Accuracy.append(comparisondf.iloc[1,2] )
Accuracy.append(comparisondf.iloc[10,2]) 
Accuracy.append(comparisondf.iloc[0,2]) 
Accuracy.append(comparisondf.iloc[9,2]) 
Accuracy.append(comparisondf.iloc[5,2])
Accuracy.append(comparisondf.iloc[14,2] )
Accuracy.append(DTCopt['default'])
Accuracy.append(scaledDTCopt['default'])
Accuracy.append(KNNopt['default'])
Accuracy.append(scaledKNNopt['default'])
Accuracy.append(LRopt['default'])
Accuracy.append(scaledLRopt['default'])
Accuracy.append(MLPCopt['default'])
Accuracy.append(scaledMLPCopt['default'])
Model = []
Model.append(comparisondf.iloc[6,1] )
Model.append(comparisondf.iloc[15,1] )
Model.append(comparisondf.iloc[1,1] )
Model.append(comparisondf.iloc[10,1]) 
Model.append(comparisondf.iloc[0,1]) 
Model.append(comparisondf.iloc[9,1]) 
Model.append(comparisondf.iloc[5,1])
Model.append(comparisondf.iloc[14,1] )
Model.append(comparisondf.iloc[6,1] )
Model.append(comparisondf.iloc[15,1] )
Model.append(comparisondf.iloc[1,1] )
Model.append(comparisondf.iloc[10,1]) 
Model.append(comparisondf.iloc[0,1]) 
Model.append(comparisondf.iloc[9,1]) 
Model.append(comparisondf.iloc[5,1])
Model.append(comparisondf.iloc[14,1] )
Scaled = []
Scaled.append('False')
Scaled.append('True')
Scaled.append('False')
Scaled.append('True') 
Scaled.append('False') 
Scaled.append('True') 
Scaled.append('False')
Scaled.append('True')
Scaled.append('False')
Scaled.append('True')
Scaled.append('False')
Scaled.append('True')
Scaled.append('False')
Scaled.append('True')
Scaled.append('False')
Scaled.append('True')
Optimized = []
Optimized.append('True')
Optimized.append('True')
Optimized.append('True')
Optimized.append('True') 
Optimized.append('True') 
Optimized.append('True') 
Optimized.append('True')
Optimized.append('True')
Optimized.append('False')
Optimized.append('False')
Optimized.append('False')
Optimized.append('False')
Optimized.append('False')
Optimized.append('False')
Optimized.append('False')
Optimized.append('False')
results_full = pd.DataFrame(list(zip(Model,Scaled,Optimized,Accuracy)), columns =['Model','Scaled','Optimized','Accuracy']) 
results_full

In [ ]:
import seaborn as sns
%matplotlib inline
g = sns.catplot(x="Accuracy", y= 'Optimized',hue="Scaled", col="Model", data=results_full, kind="bar", height=5,aspect=0.5,palette="pastel")
g.set(xlim=(0.5,0.7))

g.savefig('scaled_vs_optimizaion.png')

# Proof of concept

Para efeitos de "proof of concept" e uso atual do projecto, vamos usar o modelo com melhores resultados para prever os resultados dos jogos efectuados entre a última quinta-feira e este sábado

In [ ]:
#usar uma função de selecção, neste caso própria para problemas de classificação, 
#de modo a verificar quais as features com maior importância para o targe(win)
from sklearn.feature_selection import RFE
from sklearn.feature_selection import mutual_info_classif
import numpy as np
from sklearn.pipeline import make_pipeline
from sklearn.feature_extraction import DictVectorizer
from sklearn.feature_selection import f_classif
from sklearn.feature_selection import SelectKBest
X = dummyX
y = dummyY
def select_kbest_clf(dummyX, dummyY, k=4):
    feat_selector = SelectKBest(f_classif, k=5)
    _ = feat_selector.fit(dummyX,dummyY)
    feat_scores = pd.DataFrame()
    feat_scores["F Score"] = feat_selector.scores_
    feat_scores["P Value"] = feat_selector.pvalues_
    feat_scores["Support"] = feat_selector.get_support()
    feat_scores["Attribute"] = dummyX_train.columns
    return feat_scores
feat_scores = select_kbest_clf(dummyX,dummyY)
feat_scores = feat_scores.sort_values(by = 'F Score' , axis=0, ascending=False)
#feat_scores

In [ ]:
prev = (dummyX[['teamEFG%','opptTS%','teamFG%', 'opptOrtg', 'oppt2P%', 'team2P%']])
predictionsubsetfinal = pd.concat((dummyY,prev), axis = 1)
#predictionsubsetfinal
#para efeitos desta previsão atual criamos um subset com 6 variáveis, 3 de cada uma das equipas, que tenham um feature score alto e próximo entre elas para não enviesar a previsão para uma das equipas.

In [ ]:
#Train/test split para treinar o modelo dentro deste subset
predictionsubsetX_train = predictionsubsetfinal.loc[0:int((len(predictionsubsetfinal)*0.7)),:]
predictionsubsetX_test = predictionsubsetfinal.loc[(int(len(predictionsubsetfinal)*0.7)):len(predictionsubsetfinal),:]
predictionsubsety_train = predictionsubsetfinal.loc[0:int((len(predictionsubsetfinal)*0.7)):,'Win']
predictionsubsety_test = predictionsubsetfinal.loc[(int(len(predictionsubsetfinal)*0.7)):,'Win']
predictionsubsetX_train = predictionsubsetX_train.drop('Win', axis = 1)
predictionsubsetX_test = predictionsubsetX_test.drop('Win', axis = 1)

In [ ]:
#feat_scores entre as 6 variáveis no subset menor
from sklearn.feature_selection import RFE
from sklearn.feature_selection import mutual_info_classif
import numpy as np
from sklearn.feature_selection import f_classif
from sklearn.feature_selection import SelectKBest
X = predictionsubsetX_train
y = predictionsubsety_train
def select_kbest_clf(X, y, k=6):
    feat_selector = SelectKBest(f_classif, k=6
    _ = feat_selector.fit(X,y)
    feat_scores = pd.DataFrame()
    feat_scores["F Score"] = feat_selector.scores_
    feat_scores["P Value"] = feat_selector.pvalues_
    feat_scores["Support"] = feat_selector.get_support()
    feat_scores["Attribute"] = predictionsubsetX_train.columns
    return feat_scores
feat_scores = select_kbest_clf(X,y)
feat_scores = feat_scores.sort_values(by = 'F Score' , axis=0, ascending=False)
feat_scores

In [ ]:
#optimização dos parâmetros do modelo para o dataset agora em uso.
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics
from sklearn.metrics import accuracy_score
n_est = [2,3,4,5,6]
boot = [True,False]
crit = ['gini','entropy']
max_f = ['sqrt','log2', 'none']
max_d = [2,4,8,16,33]
RFCopt = {}
for c in crit:
    for n in n_est:
        for md in max_d:
            for b in boot:
                for mf in max_f:
                    if mf == 'none':
                        RFC = RandomForestClassifier(criterion=c,n_estimators=n,bootstrap=b, max_depth=md, random_state=0).fit(predictionsubsetX_train,predictionsubsety_train)
                    else:
                        RFC = RandomForestClassifier(criterion=c,n_estimators=n,bootstrap=b,max_features=mf, max_depth=md, random_state=0).fit(predictionsubsetX_train,predictionsubsety_train)
                    RFCpred = RFC.predict(predictionsubsetX_test)
                    comb = c+str(n)+str(b)+mf+str(md)
                    RFCopt[comb] = metrics.accuracy_score(predictionsubsety_test, RFCpred )

RFCopt
max(RFCopt, key=RFCopt.get)

In [ ]:
#treino do modelo para um dataframe apenas com estas 6 categorias estudadas, após optimização acima efectuada.
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics
from sklearn.metrics import accuracy_score
RFC = RandomForestClassifier(criterion='gini',n_estimators=6,bootstrap=False,max_features='sqrt', max_depth=2,random_state=0).fit(predictionsubsetX_train, predictionsubsety_train)
RFCpred = RFC.predict(predictionsubsetX_test)
print(metrics.accuracy_score(predictionsubsety_test, RFCpred))

In [ ]:
#previsão dos resultados de jogos de quinta-feira dia 10 a sábado dia 14.
from sklearn.ensemble import RandomForestClassifier
Previsões = pd.read_csv("C:\\Users\\Homem Das Cavernas\\Desktop\\IDS\\DatascienceProject\\Previsões.csv")
RFC = RandomForestClassifier(criterion='gini',n_estimators=6,bootstrap=False,max_features='sqrt', max_depth=2,random_state=0).fit(predictionsubsetX_train, predictionsubsety_train)
RFCpred = RFC.predict(Previsões)
RFCprob = RFC.predict_proba(Previsões)
res = [0,0,1,1,1,0,0,0,1,0,1,0,0,0,0,1,0,0]
apostas = [0,0,0,1,1,1,1,0,1,0,1,0,0,0,0,1,0,0,1,1,0,1,1,1,1,1]
Jogo = ['ORL1 vs LAL0', 'MIN1 vs UTAH0', 'PHX1 vs MEM0', 'MIL1 vs NOLA0', 'CHI1 vs ATL0', 'BOS1vs PHI0' , 'SAS1 vs CLE0','DET1 vs DAL0', 'DEN1 vs POR0', 'ORL1 vs HOU0', 'PHI1 vs NOLA0', 'MIA1 vs LAL0', 'ATL1 vs IND0', 'CHI1 vs CHA0', 'MEM1 vs MIL0', 'UTAH1 vs GSW0', 'MIN1 vs LAC0', 'SAC1 vs NYK', 'PHX1 vs SAS0', 'TOR1 vs BKN0', 'CHI1 vs LAC0', 'MEM1 vs WAS0', 'DAL1 vs MIA0', 'MIL1 vs CLE0', 'HOU1 vs DET0', 'DEN1 vs OKC0']
TabelaPrevisõesProb = pd.DataFrame(RFCprob, columns = ['AwayTeamWinProbability','HomeTeamWinTeamProbability'])
TabelaPrevisõesPred = pd.DataFrame(RFCpred, columns = ['Previsão'])
TabelaApostas = pd.DataFrame(apostas, columns = ['CasasDeApostas'])
TabelaPrevisõesResultado = pd.DataFrame(res, columns = ['Resultado Real'])
TabelaPrevisõesJogo = pd.DataFrame(Jogo, columns = ['Jogo'])
TabelaPrevisõesFinalista = [TabelaPrevisõesJogo,TabelaPrevisõesResultado,TabelaPrevisõesPred,TabelaApostas,TabelaPrevisõesProb]
TabelaPrevisõesFinal = pd.concat(TabelaPrevisõesFinalista, axis=1)
TabelaPrevisõesFinal.index = TabelaPrevisõesFinal.index + 1
TabelaPrevisõesFinal
#print(metrics.accuracy_score(res,RFCpred))
#print(RFCpred)
#print(RFCprob)

## Evaluation pipeline

## Results

## Future Prespectives

## References

In [ ]:
modelcomparison